src/containerapp/azext_containerapp/_ssh_utils.py

```diff
@@ -0,0 +1,171 @@
+
+logger = get_logger(__name__)
+
+# SSH control byte values for container app proxy
+SSH_PROXY_FORWARD = 0
+SSH_PROXY_INFO = 1
+SSH_PROXY_ERROR = 2
+
+# SSH control byte values for container app cluster
+SSH_CLUSTER_STDIN = 0
+SSH_CLUSTER_STDOUT = 1
+SSH_CLUSTER_STDERR = 2
+
+# forward byte + stdin byte
+SSH_INPUT_PREFIX = b"\x00\x00"
+
+# forward byte + terminal resize byte
+SSH_TERM_RESIZE_PREFIX = b"\x00\x04"
+
+SSH_DEFAULT_ENCODING = "utf-8"
+SSH_BACKUP_ENCODING = "latin_1"
+
+SSH_CTRL_C_MSG = b"\x00\x00\x03"
+
+
+class WebSocketConnection:
+    def __init__(self, cmd, resource_group_name, name, revision, replica, container, startup_command):
+        self._url = self._get_url(cmd=cmd, resource_group_name=resource_group_name, name=name, revision=revision,
+                                  replica=replica, container=container, startup_command=startup_command)
+        self._socket = websocket.WebSocket(enable_multithread=True)
+        logger.warning("Attempting to connect to %s", self._remove_token(self._url))
+
+        # TODO maybe catch websocket._exceptions.WebSocketBadStatusException: Handshake status 404 Not Found
+        self._socket.connect(self._url)
+        self.is_connected = True
+
+    @classmethod
+    def _remove_token(cls, url):
+        if "?token=" in url:
+            return url[:url.index("?token=")]
+        return url
+
+    @classmethod
+    def _get_url(cls, cmd, resource_group_name, name, revision, replica, container, startup_command):
+        sub = get_subscription_id(cmd.cli_ctx)
+        token_response = ContainerAppClient.get_auth_token(cmd, resource_group_name, name)
+        token = token_response["properties"]["token"]
+        logstream_endpoint = token_response["properties"]["logStreamEndpoint"]
+        proxy_api_url = logstream_endpoint[:logstream_endpoint.index("/subscriptions/")].replace("https://", "")
+
+        return (f"wss://{proxy_api_url}/subscriptions/{sub}/resourceGroups/{resource_group_name}/containerApps/{name}"
+                f"/revisions/{revision}/replicas/{replica}/containers/{container}/exec/{startup_command}?token={token}")
+
+    def disconnect(self):
+        logger.warning("Disconnecting...")
+        self.is_connected = False
+        self._socket.close()
+
+    def send(self, *args, **kwargs):
+        return self._socket.send(*args, **kwargs)
+
+    def recv(self, *args, **kwargs):
+        return self._socket.recv(*args, **kwargs)
+
+
+def _decode_and_output_to_terminal(connection: WebSocketConnection, response, encodings):
+    for i, encoding in enumerate(encodings):
+        try:
+            print(response[2:].decode(encoding), end="", flush=True)
+            break
+        except UnicodeDecodeError as e:
+            if i == len(encodings) - 1:  # ran out of encodings to try
+                connection.disconnect()
+                logger.info("Proxy Control Byte: %s", response[0])
```

Comment:
assuming logge.info is intentional over here - since these won't be visible to the user until using --verbose / --debug

Reply:
Yes I thought it would be an excessive amount of logging to show automatically in this case

Reply:
The ones I intend to display without `--debug` are done with `logger.warning` (like the direction to use `ctrl + D` to exit)